archive/imo/imo2021_q1.lean

```diff
@@ -0,0 +1,252 @@
+  ring_exp at h₁,
+  norm_num at h₁,
+  norm_cast at h₁,
+  linarith,
+end
+
+lemma upper_bound (n l : ℕ) (hl₂ : (l : ℝ) ≤ sqrt (1 + n) - 1) :
+  2 * l * l + 4 * l ≤ 2 * n :=
+begin
+  have h₁ : (l : ℝ) + 1 ≤ sqrt (1 + n), { linarith },
+  rw le_sqrt at h₁,
+  ring_exp at h₁, norm_num at h₁,
+  norm_cast at h₁,
+  linarith,
+  all_goals { norm_cast, linarith },
+end
+
+
+lemma radical_inequality {n : ℕ} (h : n ≥ 107) : sqrt (4 + 2 * n) ≤ 2 * (sqrt (1 + n) - 3) :=
+begin
+  rw sqrt_le_iff,
+  split,
+  { norm_num,
+    rw le_sqrt,
+    all_goals { norm_cast, linarith }},
+  { ring_exp,
+    rw [pow_two, ← sqrt_mul, sqrt_mul_self],
+    suffices : 24 * sqrt (1 + n) ≤ 2 * n + 36,
+    { linarith },
+    rw mul_self_le_mul_self_iff,
+    ring_exp,
+    rw [pow_two, ← sqrt_mul, sqrt_mul_self],
+    --Not splitting into cases lead to a deterministic timeout on my machine
+    by_cases p: n ≥ 108,
+    norm_cast,
+    nlinarith,
+    have : n = 107, { linarith },
+    subst this,
+    norm_num,
+    swap 3,
+    { norm_num,
+      apply sqrt_nonneg },
+    all_goals { norm_cast, linarith }},
+end
+
+/- We will later make use of the fact that there exists (l : ℕ) such that
+n ≤ 2 * l * l - 4 * l and 2 * l * l + 4 * l ≤ 2 * n for n ≥ 107. -/
+lemma exists_numbers_in_interval (n : ℕ) (hn : n ≥ 107): ∃ (l : ℕ),
+  (n + 4 * l ≤ 2 * l * l ∧ 2 * l * l + 4 * l ≤ 2 * n) :=
+begin
+  suffices : ∃ (l : ℕ), 2 * (l : ℝ) ≥ 2 + sqrt (4 + 2 * n) ∧ (l : ℝ) ≤ sqrt (1 + n) - 1,
+  { cases this with l t,
+    use l,
+    refine ⟨lower_bound n l t.1, upper_bound n l t.2⟩ },
+  lift floor (sqrt (1 + n) - 1) to ℕ with x,
+  have hx : (x : ℝ) = ⌊sqrt (1 + ↑n) - 1⌋,
+    { norm_cast,
+      convert h,
+      push_cast,
+      norm_num },
+  refine ⟨x, _, _⟩,
+  { suffices : 2 * (sqrt (1 + n) - 2) ≥ 2 + sqrt (4 + 2 * n),
+    { apply le_trans this _,
+      simp only [mul_le_mul_left, zero_lt_bit0, zero_lt_one],
+      rw hx,
+      suffices : sqrt (1 + n) - 1 ≤ ↑⌊sqrt (1 + n) - 1⌋ + 1,
+      { linarith },
+      have t :  (⌈sqrt (1 + n) - 1⌉:ℝ) ≤ ⌊sqrt (1 + n) - 1⌋ + 1,
+        norm_cast,
+        exact ceil_le_floor_add_one _,
+      apply le_trans _ t,
+      exact le_ceil (sqrt (1 + ↑n) - 1) },
+      suffices : sqrt (4 + 2 * n) ≤ 2 * (sqrt (1 + n) - 3),
+      { linarith },
+    exact radical_inequality hn },
+  { rw hx,
+    exact floor_le _ },
+  { rw [floor_nonneg, le_sub, sub_zero, le_sqrt],
+    all_goals { norm_cast, linarith }},
+
+end
+
+lemma exists_triplet_summing_to_squares (n : ℕ) (hn : n ≥ 100): (∃ (a b c : ℕ),
+  n ≤ a ∧ a < b ∧ b < c ∧ c ≤ 2 * n ∧  (∃ (k : ℕ), a + b = k * k) ∧
+  (∃ (l : ℕ), c + a = l * l) ∧ (∃ (m : ℕ), b + c = m * m)) :=
+begin
+  --If n ≥ 107, we do not explicitly construct the triplet but use an existence
+  --argument from lemma above.
+  by_cases p : n ≥ 107,
+  { have h := exists_numbers_in_interval n p,
+    cases h with l hl,
+    use [2 * l * l - 4 * l, 2 * l * l + 1, 2 * l * l + 4 * l],
+    by_cases p : l > 1,
```

Comment:
Whenever you do cases, you need to use braces to delimit the proofs of the alternative cases. (Or use the `case` tactic to state explicitly which case you're doing, but that's overkilll here.)